src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs

```diff
@@ -0,0 +1,773 @@
+                    _itemsInBatch++;
+                }
+
+                return counts;
+            }
+            finally
+            {
+                await CleanActionList();
+            }
+        }
+
+        protected override async Task<SmugglerProgressBase.Counts> ProcessLegacyAttachmentDeletionsAsync(SmugglerResult result)
+        {
+            var counts = new SmugglerProgressBase.Counts();
+            try
+            {
+                foreach (var des in _destinations)
+                {
+                    _actionsList.Add(des.LegacyAttachmentDeletions());
+                }
+                await foreach (var id in _source.GetLegacyAttachmentDeletionsAsync())
+                {
+                    HandleImportBatch("LegacyAttachmentDeletions");
+                    var index = ShardHelper.GetShardIndexforDocument(_transactionOperationContext, _shardAllocation, id);
+                    await ((ILegacyActions)_actionsList[index]).WriteLegacyDeletions(id);
+                    _itemsInBatch++;
+                }
+
+                return counts;
+            }
+            finally
+            {
+                await CleanActionList();
+            }
+        }
+
+        protected override async Task<SmugglerProgressBase.Counts> ProcessCompareExchangeAsync(SmugglerResult result)
+        {
+            result.CompareExchange.Start();
+
+            foreach (var des in _destinations)
+            {
+                _actionsList.Add(des.CompareExchange(_context));
+            }
+
+            await using (var actions = new DatabaseCompareExchangeActions(_server, _databaseRecord, _context, new CancellationToken()))
+            {
+                try
+                {
+                    await foreach (var kvp in _source.GetCompareExchangeValuesAsync())
+                    {
+                        _token.ThrowIfCancellationRequested();
+                        result.CompareExchange.ReadCount++;
+                        if (result.CompareExchange.ReadCount != 0 && result.CompareExchange.ReadCount % 1000 == 0)
+                            AddInfoToSmugglerResult(result, $"Read {result.CompareExchange.ReadCount:#,#;;0} compare exchange values.");
+
+                        if (kvp.Equals(default))
+                        {
+                            result.CompareExchange.ErroredCount++;
+                            continue;
+                        }
+
+                        try
+                        {
+                            if (ClusterTransactionCommand.IsAtomicGuardKey(kvp.Key.Key, out var docId))
+                            {
+                                HandleImportBatch(nameof(DatabaseItemType.CompareExchange));
+                                var index = ShardHelper.GetShardIndexforDocument(_transactionOperationContext, _shardAllocation, docId);
+                                await ((ICompareExchangeActions)_actionsList[index]).WriteKeyValueAsync(kvp.Key.Key, kvp.Value);
```

Comment:
Aren't they on the same "key space", so to speak? They all share the same cmpxng space?